### PR TITLE
Xds proxy improve

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
This solves two main issues:

1. prevent upstream receive goroutine from leaking
2. split upstream request handling and response handling, as we know the send could take very long

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
cc @howardjohn @ramaraochavali 

This is an alternative

**Comment by ramaraochavali** [comment]
This seems a better alternative - I have not looked at in detail yet but It seems if either stream is disconnected you are reconnecting both?

And by removing cert watches we are assuming that till next disconnect (however long it could be), the connection uses the old certs (which might have expired) - Is that correct?

File: pkg/istio-agent/xds_proxy.go [hunk 1]
```diff
@@ -572,3 +533,21 @@ func sendUpstreamWithTimeout(ctx context.Context, upstream discovery.AggregatedD
 		return err
 	}
 }
+
+// sendDownstreamWithTimeout sends discovery response with default send timeout.
+func sendDownstreamWithTimeout(downstream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer,
```
Comment (howardjohn) [inline pkg/istio-agent/xds_proxy.go:538]: Does envoy have a timeout for send/recv on xds?

**Comment by hzxuzhonghu** [comment]
>This seems a better alternative - I have not looked at in detail yet but It seems if either stream is disconnected you are reconnecting both?

Yes, this is unchanged. As you both said this is to make xds-proxy a dumb proxy.

>And by removing cert watches we are assuming that till next disconnect (however long it could be), the connection uses the old certs (which might have expired) - Is that correct?

Yes, as you can see the reset channel is unused before this.

File: pkg/istio-agent/xds_proxy.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:538]: I am not sure, let me have a check

**Comment by howardjohn** [comment]
btw, looks good to me. just didn't approve since wanted to make sure @ramaraochavali was ok with the cert changes

File: pkg/istio-agent/xds_proxy.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:538]: Normally i think for receive, it should be no.

**Comment by hzxuzhonghu** [comment]
As for tls cert related issue: after a connection established with a certificate, the c/s communicate with each other by encrypting the data with negotiated symmetric encryption.  IIUC, this has nothing to do with the certificate now.

File: pkg/istio-agent/xds_proxy.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:541]: It seems envoy is using async client to handle xds stream.

```cpp
void GrpcMuxImpl::establishNewStream() {
  ENVOY_LOG(debug, "Establishing new gRPC bidi stream for {}", service_method_.DebugString());
  stream_ = async_client_->start(service_method_, *this);
  if (stream_ == nullptr) {
    ENVOY_LOG(warn, "Unable to establish new stream");
    handleFailure();
    return;
  }

  control_plane_stats_.connected_state_.set(1);
  for (const auto type_url : subscriptions_) {
    sendDiscoveryRequest(type_url);
  }
}
```

Comment (ramaraochavali) [reply pkg/istio-agent/xds_proxy.go:541]: I do not see Envoy setting timeout for XDS requests

**Review by ramaraochavali — APPROVED** [review]
LGTM.  Reg cert watches, It may be OK because we are reestablishing the connection after reset (not pooling it).

**Pull request merged**